src/main/java/com/google/devtools/build/lib/windows/WindowsFileSystem.java

```diff
@@ -92,7 +92,11 @@ protected void createSymbolicLink(Path linkPath, PathFragment targetFragment) th
   @Override
   protected PathFragment readSymbolicLink(Path path) throws IOException {
     java.nio.file.Path nioPath = getNioPath(path);
-    return PathFragment.create(WindowsFileOperations.readSymlinkOrJunction(nioPath.toString()));
+    try {
+      return PathFragment.create(WindowsFileOperations.readSymlinkOrJunction(nioPath.toString()));
+    } catch (IOException e) {
+      throw e.getMessage().endsWith("path is not a link") ? new NotASymlinkException(path) : e;
```

Comment:
Don't branch on the exception message, let `WindowsFileOperations.readSymlinkOrJunction` throw the right exception instead.

Reply:
WindowsFileOperations.readSymlinkOrJunction doesn't have access to NotASymlinkException (it's low-level jni wrapper). 
Similarly (I suspect that for the same reason) UnixFileSystem.readSymbolicLink branches on the exception message.
I would leave it the way it is.

Reply:
Please change `WindowsFileOperations.readSymlinkOrJunction` to return an error enum then, instead of throwing `IOException`. `WindowsFileOperations` is Bazel's own internal class so we can change it freely.

`UnixFileSystem.readSymbolicLink` sets a bad example: text output meant for humans should not be an API contract because it's too easy to break without noticing. (I know there's code in Bazel that violates this because of [Hyrum's law](https://www.hyrumslaw.com), but that's bad and we shouldn't introduce more if possible.)